### PR TITLE
Add support for Access Token Authentication for SQL Server Driver (mssql)

### PR DESCRIPTION
**Pull request by funkydev**
### Overview
When I was modifying `mmsql` package, I noticed that it accepts `authentication` parameters in the config object. See discussion under PR https://github.com/tediousjs/node-mssql/pull/1208 and documentation I've updated https://github.com/tediousjs/node-mssql#tedious. It passes authentication configuration directly to the `tedious` library, which supports the following authentication types.

- `DefaultAuthentication`
- `NtlmAuthentication`
- `AzureActiveDirectoryAccessTokenAuthentication`
- `AzureActiveDirectoryMsiAppServiceAuthentication`
- `AzureActiveDirectoryMsiVmAuthentication`
- `AzureActiveDirectoryPasswordAuthentication`
- `AzureActiveDirectoryServicePrincipalSecret`

I copied authentication options interfaces directly from the `tedious` repository and modified TypeORM code to support passing authentication options for the sqlserver driver.

### What does it change?
It changes a way of passing credentials to the SQLServer database. I modified current interfaces to deprecate `domain` property directly in the configuration model. Instead of this I proposed to pass `authentication` property which is compatible with the `mssql` and `tedious` one. 

I'd recommend to remove `options.domain` in future, to give developers time for any adjustments.  Then instead of passing: 
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  username: "username",
  password: "password",
  domain: "domain"
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```
developers should pass:
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "ntlm",
    options: {
        userName: "username",
        password: "password",
        domain: "domain"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

### Solves
- https://github.com/typeorm/typeorm/issues/4510 
- https://github.com/typeorm/typeorm/issues/7340

#### Default authentication type
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "default",
    options: {
      userName: "username",
      password: "password"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

#### AAD Access Token
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "azure-active-directory-access-token",
    options: {
      token: "token"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

#### AAD Service Principal (secret)
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "azure-active-directory-service-principal-secret",
    options: {
      tenantId: 'tenant-id',
      clientId: 'client-id',
      clientSecret: 'client-secret'
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

#### AAD MSI App Service Authentication
```javascript
createConnection({
  type: "mssql",
  database: "db-name",
  host: "db-server",
  authentication: {
    type: "azure-active-directory-msi-app-service",
    options: {
      clientId: "optional clientId (it should be taken from app service environment)",
      msiEndpoint: "optional msiEndpoint (it should be taken from app service environment)",
      msiSecret: "optional msiSecret (it should be taken from app service environment)"
    }
  }
}).then(connection => {
  // here you can start to work with your entities
}).catch(error => console.log(error));
```

### CODE REVIEW

**Comment by funkydev** [comment]
@mnotjames Could you check this one?

**Comment by funkydev** [comment]
@imnotjames Could you check this one?

File: src/driver/sqlserver/authentication/AzureActiveDirectoryPasswordAuthentication.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+export interface AzureActiveDirectoryPasswordAuthentication {
+    type: "azure-active-directory-password";
+    options: {
+        /**
+         * A user need to provide `userName` asscoiate to their account.
+         */
+        userName: string;
```
Comment (pleerock) [inline src/driver/sqlserver/authentication/AzureActiveDirectoryPasswordAuthentication.ts:7]: we have `username` everywhere written in lowercase. Can we do same here?

**Comment by AlexMesser** [comment]
thank you for contribution!

**Pull request merged**